src/file-upload/internal.tsx

```diff
@@ -0,0 +1,181 @@
+
+  if (!multiple && value.length > 1) {
+    warnOnce('FileUpload', 'Value must be an array of size 0 or 1 when `multiple=false`.');
+  }
+
+  const handleFilesChange = (newFiles: File[]) => {
+    const newValue = multiple ? [...value, ...newFiles] : newFiles[0] ? newFiles : [...value];
+    fireNonCancelableEvent(onChange, { value: newValue });
+  };
+
+  const onFileRemove = (removeFileIndex: number) => {
+    const newValue = value.filter((_, fileIndex) => fileIndex !== removeFileIndex);
+    fireNonCancelableEvent(onChange, { value: newValue });
+    setRemovedFileIndex(removeFileIndex);
+    setRemovedLastFile(value.length === 1);
+  };
+
+  useEffect(() => {
+    if (removedLastFile) {
+      fileInputRef.current?.focus();
+    }
+  }, [removedLastFile]);
+
+  const isDropzoneVisible = useDropzoneVisible();
+
+  const formFieldContext = useFormFieldContext(restProps);
+  const ariaDescribedBy = joinStrings(
+    restProps.ariaDescribedby ?? formFieldContext.ariaDescribedby,
+    errorText ? errorId : undefined,
+    constraintText ? constraintTextId : undefined
+  );
+
+  const hasError = Boolean(errorText || fileErrors?.filter(Boolean).length);
+  const invalid = restProps.invalid || formFieldContext.invalid || hasError;
+
+  return (
+    <InternalSpaceBetween
+      {...baseProps}
+      size="xs"
+      className={clsx(baseProps.className, styles.root)}
+      __internalRootRef={__internalRootRef}
+    >
+      {isDropzoneVisible ? (
+        <Dropzone onChange={handleFilesChange}>{i18nStrings.dropzoneText(multiple)}</Dropzone>
+      ) : (
+        <FileInput
+          ref={ref}
+          accept={accept}
+          ariaRequired={ariaRequired}
+          multiple={multiple}
+          onChange={handleFilesChange}
+          value={value}
+          {...restProps}
+          ariaDescribedby={ariaDescribedBy}
+          invalid={invalid}
+        >
+          {i18nStrings.uploadButtonText(multiple)}
+        </FileInput>
+      )}
+
+      {(constraintText || errorText) && (
+        <div className={styles.hints}>
+          {errorText && (
+            <FormFieldError id={errorId} errorIconAriaLabel={i18nStrings?.errorIconAriaLabel}>
+              {errorText}
+            </FormFieldError>
+          )}
+          {constraintText && (
+            <ConstraintText id={constraintTextId} hasError={!!errorText}>
+              {constraintText}
+            </ConstraintText>
+          )}
+        </div>
+      )}
+
+      {!multiple && value.length === 1 ? (
+        <Token
+          ariaLabel={value[0].name}
+          dismissLabel={i18nStrings.removeFileAriaLabel(value[0], 0)}
+          onDismiss={() => onFileRemove(0)}
+          errorText={fileErrors?.[0]}
+          errorIconAriaLabel={i18nStrings.errorIconAriaLabel}
+        >
+          <FileOption file={value[0]} metadata={metadata} i18nStrings={i18nStrings} />
+        </Token>
+      ) : null}
+
+      {(multiple && value.length > 0) || value.length > 1 ? (
```

Comment:
When is the `value.length > 1` condition needed?

Reply:
That is because technically one can set `multiple=false` and still provide multiple files or even change `multiple` dynamically. In that case we issue a dev. warning but also there needs to be some representation. One option would be to show the first file only but I think a better fair one is to show all files.

Reply:
In table single selection mode we simply ignore extra items: https://github.com/cloudscape-design/components/blob/main/src/table/use-selection.ts#L97

I suggest to not over-engineer here too, and not support invalid use-cases

Reply:
Alright, I am not against that